### ISSUE TITLE
Deconstruction in foreach does not appropriately infer nullable types

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#37491]
```cs
class C
{
    void M(object? o1)
    {
        foreach ((var o2, object o3) in new[] { (o1, o1) })
        { 
            o3.ToString();
        }
    }
}
```

The call to `o3.ToString()` should warn, and there should be a warning on converting a nullable reference type to a non-nullable reference type in the assignment to o3.
